lightclient/src/platform/default.rs

```diff
@@ -0,0 +1,299 @@
+    type Stream = platform_impl::Stream;
+    type ConnectFuture = future::BoxFuture<
+        'static,
+        Result<PlatformConnection<Self::Stream, Self::Connection>, ConnectError>,
+    >;
+    type StreamUpdateFuture<'a> = future::BoxFuture<'a, ()>;
+    type NextSubstreamFuture<'a> =
+        future::Pending<Option<(Self::Stream, PlatformSubstreamDirection)>>;
+
+    fn now_from_unix_epoch(&self) -> Duration {
+        platform_impl::now_from_unix_epoch()
+    }
+
+    fn now(&self) -> Self::Instant {
+        platform_impl::now()
+    }
+
+    fn sleep(&self, duration: Duration) -> Self::Delay {
+        platform_impl::sleep(duration)
+    }
+
+    fn sleep_until(&self, when: Self::Instant) -> Self::Delay {
+        self.sleep(when.saturating_duration_since(self.now()))
+    }
+
+    fn yield_after_cpu_intensive(&self) -> Self::Yield {
+        // No-op.
+        future::ready(())
+    }
+
+    fn connect(&self, multiaddr: &str) -> Self::ConnectFuture {
+        // We simply copy the address to own it. We could be more zero-cost here, but doing so
+        // would considerably complicate the implementation.
+        let multiaddr = multiaddr.to_owned();
+
+        tracing::debug!("Connecting to multiaddress={:?}", multiaddr);
+
+        Box::pin(async move {
+            let addr = multiaddr.parse::<Multiaddr>().map_err(|_| ConnectError {
+                is_bad_addr: true,
+                message: "Failed to parse address".to_string(),
+            })?;
+
+            let mut iter = addr.iter().fuse();
+            let proto1 = iter.next().ok_or(ConnectError {
+                is_bad_addr: true,
+                message: "Unknown protocols combination".to_string(),
+            })?;
+            let proto2 = iter.next().ok_or(ConnectError {
+                is_bad_addr: true,
+                message: "Unknown protocols combination".to_string(),
+            })?;
+            let proto3 = iter.next();
+
+            if iter.next().is_some() {
+                return Err(ConnectError {
+                    is_bad_addr: true,
+                    message: "Unknown protocols combination".to_string(),
+                });
+            }
+
+            platform_impl::connect(proto1, proto2, proto3).await
+        })
+    }
+
+    fn open_out_substream(&self, c: &mut Self::Connection) {
+        // This function can only be called with so-called "multi-stream" connections. We never
+        // open such connection.
+        match *c {}
+    }
+
+    fn next_substream<'a>(&self, c: &'a mut Self::Connection) -> Self::NextSubstreamFuture<'a> {
+        // This function can only be called with so-called "multi-stream" connections. We never
+        // open such connection.
+        match *c {}
+    }
+
+    fn update_stream<'a>(&self, stream: &'a mut Self::Stream) -> Self::StreamUpdateFuture<'a> {
+        Box::pin(future::poll_fn(|cx| {
+            let Some((read_buffer, write_buffer)) = stream.buffers.as_mut() else { return Poll::Pending };
```

Comment:
`Poll::Pending` here would get stuck forever (nothing would ever wake it up again except by chance); does it make more sense to return an error of some kind?

Reply:
Indeed, the `Poll::Pending` here will never wake up, since we don't store the context anywhere. 

The stream's buffers can only be empty if the connection is in the `Reset` state. The `fn connect(&self, multiaddr: &str) -> Self::ConnectFuture ` method is called first to establish the connection that initializes the buffers here.

Looking a bit at smoldot, my understanding is:
- A task is spawned to handle incoming connection attempts ([here](https://github.com/smol-dot/smoldot/blob/5bb664978d0576fe4d19b9ed6f7537798d238e43/light-base/src/network_service.rs#L1058-L1068))
- The task checks in a loop that progress can be made on the connect; and updates some internal state machine
- The loop is ended when it cannot send a message back to the background task; or when the socket is closed 
- The last step of that loop waits for one of the following to happen
  - stream update (this future) to complete
  - receive a message from the coordinator (some internal detail of smoldot)
  - a timeout to happen
  
I believe that we drop this future after some period of time when we return `Poll::Pending` here.

The smoldot's `type StreamUpdateFuture<'a>: Future<Output = ()> + Unpin + Send + 'a;` forces us to return `()` and considering the above we might introduce a busy wait loop if we could return `Poll::Ready(Err()))`

// cc @tomaka would love to hear your thoughts on this 🙏

Reply:
You don't need to read smoldot's source code, you just have to read the documentation of the function you're implementing: https://docs.rs/smoldot-light/latest/smoldot_light/platform/trait.PlatformRef.html#tymethod.update_stream

It says:

> Returns a future that becomes ready when “something” in the state has changed.

Once a connection is in the reset state, it state never changes, and thus returning an infinite future is correct.

Reply:
The reason why the connections-related functions in the `Platform` trait are so different from `AsyncRead` and `AsyncWrite` is that `AsyncRead` and `AsyncWrite` are full of weird corner cases.

What errors can actually be returned and what do they represent? For example, can you suddely get a `PermissionDenied` error?
Can the reading side return errors and the writing side still work, or vice versa?
Are some errors recoverable? Can you continue reading/writing again despite an error being returned?
If you call `poll_close` and it returns `Pending`, can you start calling `poll_write` again after?

So to solve all of this I created my own abstraction which follows what I think are straight forward rules. There are no errors, and just 5 possibles states (open, read open write closed, read closed write open, both closed, and reset).
It's a bit annoying to plug into existing APIs, but I'm going to provide helpers in the future.